File: src/main/java/io/github/slavsquatsuperstar/mayonez/input/KeyAxis.java
package io.github.slavsquatsuperstar.mayonez.input;

public enum KeyAxis {

	VERTICAL(KeyMapping.DOWN, KeyMapping.UP), HORIZONTAL(KeyMapping.RIGHT, KeyMapping.LEFT);

	private int value;
	private int posKeyCode, negKeyCode;
	private boolean posDown, negDown;

	// make reference to key instead of code
	private KeyAxis(KeyMapping posKey, KeyMapping negKey) {
		this.posKeyCode = posKey.keyCode();
		this.negKeyCode = negKey.keyCode();
	}

	// if contains key code
	public boolean match(int keyCode) {
		return keyCode == posKeyCode || keyCode == negKeyCode;
	}

	// receive new key event
	public void setKey(int keyCode, boolean keyDown) {

		if (keyDown) { // every new input will override the old one

			if (keyCode == posKeyCode) {
				posDown = true;
				value = 1;
			} else if (keyCode == negKeyCode) {
				negDown = true;
				value = -1;
			}

		} else { // will remember which keys are down when letting go

			if (keyCode == posKeyCode) {
				posDown = false;
				value = negDown ? -1 : 0;
			} else if (keyCode == negKeyCode) {
				negDown = false;
				value = posDown ? 1 : 0;
			}

		}

	}

	public int value() {
		return value;
	}
	
	@Override
	public String toString() {
		return name().toLowerCase();
	}

}

File: src/main/kotlin/slavsquatsuperstar/mayonez/input/KeyAxis.kt
package slavsquatsuperstar.mayonez.input

import org.apache.commons.lang3.StringUtils

/**
 * Stores two keys intended to perform opposite actions.
 */
// TODO store mappings?
internal enum class KeyAxis(posKey: KeyMapping, negKey: KeyMapping) {
    VERTICAL(KeyMapping.DOWN, KeyMapping.UP), HORIZONTAL(KeyMapping.RIGHT, KeyMapping.LEFT);

    private val posKey: String = posKey.name
    private val negKey: String = negKey.name

    /**
     * @return The value of this axis. 1 if the positive key is pressed. -1 if the negative key is pressed.
     * 0 if the both or neither key is pressed.
     */
    internal fun value(): Int {
        // add neg and pos keys to make sure pressing a new key doesn't override the other
        val negVal = if (KeyInput.keyDown(negKey)) -1 else 0
        val posVal = if (KeyInput.keyDown(posKey)) 1 else 0
        return negVal + posVal
    }

    override fun toString(): String = StringUtils.capitalize(name.lowercase())

}